baybe/searchspace/_filtered.py

```diff
@@ -0,0 +1,45 @@
+"""Search spaces with metadata."""
+
+import numpy as np
+import numpy.typing as npt
+import pandas as pd
+from attrs import asdict, cmp_using, define, field
+from attrs.validators import instance_of
+from typing_extensions import Self, override
+
+from baybe.searchspace import SubspaceDiscrete
+
+
+@define
+class FilteredSubspaceDiscrete(SubspaceDiscrete):
```

Comment:
As already said, not strictly required. But I did it this way to keep the changes in the PR to a minimum for you, in a way that does not introduced backward-incompatible changes later.

I suggest:
* Let's keep it for now, to finally solve the immutability problem (stuff is completely hidden from the user)
* Have one proper PR that cleanly implements the `SubspaceDiscrete.filter` logic, with all related code improvements / simplifications (i.e. use the method as the main work horse for filtering, also for our current constraints). Also your suggested approach would then rely on this method, i.e. we'd simply call `filter` instead of `from_masked_subspace`

If you agree, simply resolve.

Reply:
I dont see how adding a new class is `keeping changes to a minimum`, I would much prefer it becoming part of the existing class and not adding an obsolete new class

As already clarified int he meeting, this is not related to any of the potential changes around your filtering endeavors

Reply:
Ok, I probably did not express myself clearly 🙃  What I meant: taking care of immutability while postponing the decision (in a backward-compatible way) how to handle filtering operations in detail – Because note: the mask itself IS already a filtering mechanism itself, and thus adds statefulness.

And now I also remember the main reason why I added the separate class: separating stateful code from stateless code. If you just add the mask back to the original class, you are exactly back to where we were!! The main focus here was to keep `SubspaceDiscrete` stateless and outsource the statefulness into a separate class that is hidden from the user.

So how do we proceed? I see three ways:
1. We handle this part in a separate PR to avoid having yet another review round on the existing one (which would be definitely needed)
2. I move the mask as an internal object back to the original class and make it read-only, which I think is possible for numpy arrays, keeping the statelessness.
3. I move it back and and make it private + we don't care about the fact that it is mutable because it's private. The PR would then still solve the other issue, i.e. metadata and flags being handled in the campaign. But somehow this seems like it's opening the door again to the original problem.

Reply:
hmm ok I forgot about the statefulness again, seems then OK as it is for this PR

Reply:
Ok, resolved then